UFO/OpenType conversion

dpk opened the pull request:
In 2015 I forked this repository, converted the files to OpenType format using FontForge and added several nifty OpenType features. At the time, I was modestly satisfied to have this fork for my own use without merging my (rather dramatic changes) upstream. Those who wanted these features could find them in my repo; the original remained here, untouched.

Nearly six years later, I have redone this conversion using a more modern, Git-friendly UFO-based process using Adobe’s open-source AFDKO tools, and this time, I would rather immodestly like to suggest that my version should, when I’ve completed and checked everything, be the new official version of the ET Book open source fonts.

There are a number of issues in this repository’s tracker which either have already been fixed or will be fixed in my version. #19 and some of #20 have been fixed (and as far as #19 is concerned, I intend to make further improvements on kerning later), #21 should be fixed when the fix on #20 is done, and I intend to make a start on #22/#24 soon. Further, as @skosch pointed out in #20, using a UFO-based toolchain would make this project far more contribution-friendly — at the moment, there are no font source files in this main repo which a typical font editor application could edit properly, only completed binaries converted from the original PostScript Type 1 fonts.

There is one stipulation I would add before allowing my branch to be merged into the main upstream repo: that #2 be dealt with. I cannot do this alone as I am (obviously) not the original copyright holder on the fonts, but I am willing to relicense all my own contributions under the OFL if the licence on this repository changes. The MIT licence is extremely unsuitable for fonts and its actual requirements are almost certainly not those which @tufte and his team intended to impose on users of the open-source versions of his fonts. Technically, as it stands, every web page using this font should include a copy of the MIT licence — this has not been enforced and I can’t imagine that it was intended. The OFL has some minor issue, but is the *de facto* standard for open source font licensing. I would like to repeat @davelab6’s plea and ask the copyright owners of the original fonts (Dmitry Krasny, Bonnie Scranton, and Edward Tufte) to reconsider their choice of licence and use the OFL instead. (There was also a suggestion in #17 to use the GPL with the Font Exception instead. I would possibly be open to doing that; indeed, it may be that I would prefer it to the OFL, but I’m not sure of the exact difference between the two in practice. I’ll await feedback from those more knowledgeable than me.)

This pull request is not yet ready to merge — I’ve opened it to solicit discussion about whether this should happen, and in particular to try to get issue #2 pushed along a bit.

dpk committed:
initial conversion to UFO format

dpk committed:
delete old conversion

dpk committed:
add fontplots of the UFOs to more easily see what character is which

dpk committed:
first try at merging the Roman LF into the Roman OSF font

dpk committed:
corrected support for OSF/LF in the single Roman-style font

dpk committed:
automatic ligature substitution in the new Roman font

dpk committed:
accidentally committed a test build file

dpk committed:
add ligatures feature to the Semibold font, which is now renamed to be sensible

dpk committed:
rename Bold OSF to bold ready to merge the lining numerals in

dpk committed:
merge lining figures into the Bold font

dpk committed:
add ligature support to the bold font

dpk committed:
add ligatures to italic

dpk committed:
add ligatures to display italic

dpk committed:
add the precomposed fractions from the expert font to the Roman font

dpk committed:
small caps support for letters

dpk committed:
add AFM files from Fondu (contains basic kerning information)

dpk committed:
redo Roman font kerning from the original font suitcase metrics

dpk committed:
add a README file which actually documents what's going on in this branch

dpk committed:
add pyfontaine to requirements.txt

This will help determine which characters I need to add to support
more languages.

dpk committed:
add c2sc punctuation; fixes #11

dpk committed:
add proportional numbers from the figure kerning in the expert font

dpk committed:
kerning for all the fonts

dpk committed:
new version of Makefile

dpk committed:
add WOFF and WOFF2 builds to the Makefile

skosch commented:
Wow, incredible. This is a massive gift to the community—thank you for this work :slightly_smiling_face: 

I, for one, wasn't aware this was under MIT. What an odd choice. I can only assume that the original creators were concerned only with the source files, and didn't consider the binaries to require inclusion of the license text (or perhaps that inclusion in the name table would do). @adamschwartz & @tufte please clarify—the easiest solution is to switch to OFL, which is MIT-in-spirit but tailored to fonts.

@Discordius this PR is the real deal, I'm happy to turn over my bounty to @dpk if this gets merged and solves your problem.

dpk committed:
add glyph2svgpath script that can be used to check the rendering of glpyhs

dpk committed:
add numerators, denominators, and OpenType fraction support

dpk committed:
fix style mapping for the Roman font

dpk committed:
fixes for Roman kerning (added small caps kerning)

dpk committed:
update README with news of fractions

dpk committed:
add pangrams page to test language support

dpk committed:
add doc directory

dpk committed:
begin documenting accent positions for italic

dpk commented:
Here’s a quick status update:

* Essentially all of the merging of features and characters from the original fonts into consolidated UFO files using OpenType features is done, including the fiddly fractions feature I could never get working in the old version. This means that you access old style vs lining figures using the corresponding OpenType feature (accessible, for instance, though `font-variant-numerals` in CSS, the ⌘T Typography panel in Mac OS, `Numbers=Lining` or `OldStyle` in fontspec for LaTeX, etc.) instead of loading separate lining and old-style fonts (saves bandwidth), small caps likewise through the OpenType feature instead of the Expert font, etc. There are still some different-looking characters in the Expert font which I’m not quite sure how I can sensibly merge into the main Roman font, but I’ll sort it out.
* This includes kerning information from the fonts which was missing both in my original conversion and in the one done for the upstream repo here, albeit that it’s unfortunately very limited and somewhat inconsistent. Sometime after 2.0 I’d like to investigate the possibility of funding the use of [iKern](https://www.ikern.space) on the fonts. (I got a quote last year and it was well under €1000 iirc, but I’ve since lost the email thread. D’oh.) ([Issue 12 in my fork.](https://github.com/dpk/et-book/issues/12))
* My goal for 2.0 is to support *at least* all the official languages of the European Union which use Latin script (most of the ones which aren’t supported so far are Eastern European languages). To this end I have been analysing the metrics and diacritic placements on the existing glyphs and coming up with a way to compose accented characters nicely. I hope to be able to generate a huge number of appropriate-looking accented characters automatically based on this analysis. I’ve also drawn a Croatian d-with-crossbar in all the weights and styles of the font, which was the only missing character needed for this set that isn’t a simple base character + diacritic combination. ([Issue 2 in my fork.](https://github.com/dpk/et-book/issues/2))
* Shortly after 2.0 I hope to extend that to all the minority and regional languages of Europe, too. To that end I’ve drawn eng, Eng, and ezh (for Sami, but also useful for some African languages and English IPA) in the Roman font, and am looking at the possibilities for drawing them in the other weights and styles too. (Same issue.)
* Since I was messing around in the glyph editor anyway and it was relatively easy to put together, 2.0 will also likely include an fj ligature (useful for Norwegian, Swedish, etc., and for the English loanword fjord) — like đ, I now have it ready in all weights and styles.
* The hinting issue (#20 and #21 in this repo) continues to be somewhat vexing. I’ve learned a lot about how PostScript font hinting works and how to get `psautohint` to work a bit better on the font, but there are still some behaviours I just don’t understand (most bafflingly, hinting causing characters which were drawn shorter to appear larger in comparison to other characters — and vice versa)! I tried converting the finished OTFs to TTFs and running `ttfautohint` over them, but the result didn’t look any better than having no hinting at all. `psautohint` hints more aggressively and thus gives more legible results, but also requires more work to get it not to do strange things. This is the one issue I could most use outside help with, from someone who knows how these things really work. ([Issue 4 in my fork.](https://github.com/dpk/et-book/issues/4))

And now the bad news: this has been a fun distraction between two exam seasons for me while trying to work off the stress from the first one, but I do have to study for an exam near the end of the month, and I have other (paying) work to do as well. So progress might slow a bit now. I’m good at using projects like this to procrastinate other things, though 🙈, so it definitely won’t grind completely to a halt ;-)

dpk commented:
Gaaaack. If you rename a branch in GitHub, it thinks (for the purposes of pull requests) that you deleted it. I guess I’m stuck with the `2.0` branch in my fork until this gets merged here 🤦🏼‍♀️

dpk commented:
An alternative solution to the licence problem just occurred to me. The fonts have copyright metadata (which atm I’ve left unchanged). It would satisfy the requirements of the MIT licence to put the licence text into that metadata field — as long as WOFF and WOFF2 converters (including subsetters) don’t strip that field out to save bandwidth. As long as that’s true, everyone who uses the font on the web etc. complies with the letter of the licence without having to take any further steps.

As the licence text is rather long, a link to e.g. <https://opensource.org/licenses/MIT> in this field would also loosely satisfy the requirement that the copyright and licence notice be included with the font. It would comply with the spirit of the licence, even if not the letter (and there are, of course, open-source projects (mainly smaller ones) which put themselves under a licence just by linking to the text).

Then we just have to bikeshed the exact form of the copyright string. E.g.

> Copyright (c) 2015-2021 Dmitry Krasny, Bonnie Scranton, Edward Tufte, and contributors. Licence: https://opensource.org/licenses/MIT

dpk committed:
re-add fi and fl glyphs, fixes #13

dpk committed:
fix Markdown error in accent position docs

dpk committed:
add dcroat and Dcroat to Roman and Italic fonts

dpk committed:
normalize italic dcroat, add semibold dcroat

dpk committed:
add dcroat to bold font

dpk committed:
add dcroat and Dcroat to DisplayItalic font

dpk committed:
add draft f_j and f.deu glyphs

dpk committed:
correct names of the draft f_j glyphs

dpk committed:
draft glyphs for double-dagger and single guillemets

See #17

dpk committed:
update README with latest news

dpk committed:
Add hinting test for Roman font w/ some blue values that look okay.

The blue values were detected by ttfautohint and fed back into
psautohint. A few characters still look weird at size 17 (c, g, h, s,
and a slight odd-looking blob on n).

dpk committed:
capitalization typo fix in README

dpk committed:
Generate hinted TTFs and test their hinting too.

dpk committed:
don't error on mkdir if dir already exists

dpk committed:
DRY in the Makefile.

dpk committed:
(unsuccessful) attempt to tweak blue values

dpk committed:
Make WOFFs from TTFs, not OTFs, for now, as TTF hinting is better.

dpk committed:
convert originals to AppleSingle, so they're not useless empty files

dpk commented:
My work on the font has stalled again, but it looks like the first outing for the new đ character will be in my current term paper for university — and not in Croatian but in (an older reconstruction of) Proto-Indo-European!

<img width="987" alt="Screenshot 2021-09-21 at 11 50 16" src="https://user-images.githubusercontent.com/378375/134150484-2d84240a-95ee-496a-8d56-2cbf64f0ecd7.png">

(Don’t be confused by the other funny characters — they’re mostly bodged together in TeX from the other accent marks that are available in ET Book already. The superscript w is just a superscript w, and the gamma is from GFS Porson (which works well as a drop-in to mix with the italic, don’t you find?).)

skosch commented:
Nice! I was just about to ask about the gamma – it looks great!

dpk committed:
Remove spurious <unicode hex="..." /> tags from OT variant characters

dpk committed:
Correct name of lining zero and six glyphs

dpk committed:
New draft Roman double-dagger with fixed curves

dpk committed:
ignore glyphs autosaves

dpk committed:
Italic double dagger draft glyph

dpk committed:
standardize draft glyph names

dpk committed:
draft Roman font vowels with ogonek + hbar

dpk committed:
merge fj ligatures into fonts

dpk committed:
Add single guillemets to Roman font

dpk committed:
Convert Roman to Glyphs UFO and add aogonek

dpk committed:
Aogonek

inklesspen commented:
@dpk in your opinion, is this branch currently suitable for use as a webfont for English-language content?

adamschwartz commented:
@dpk Would you recommend I merge it? You’ve done so much great work here. I’m happy to defer to you. (And thank you!)
